token-upgrade/program/tests/functional.rs

```diff
@@ -0,0 +1,305 @@
+#![cfg(feature = "test-sbf")]
+
+use {
+    solana_program_test::{
+        processor,
+        tokio::{self, sync::Mutex},
+        ProgramTest, ProgramTestContext,
+    },
+    solana_sdk::{
+        instruction::{AccountMeta, InstructionError},
+        pubkey::Pubkey,
+        signature::Signer,
+        signer::keypair::Keypair,
+        transaction::{Transaction, TransactionError},
+    },
+    spl_token_client::{
+        client::{
+            ProgramBanksClient, ProgramBanksClientProcessTransaction, ProgramClient,
+            SendTransaction,
+        },
+        token::Token,
+    },
+    spl_token_upgrade::{
+        error::TokenUpgradeError, get_token_upgrade_authority_address, instruction::exchange,
+    },
+    std::sync::Arc,
+    test_case::test_case,
+};
+
+fn keypair_clone(kp: &Keypair) -> Keypair {
+    Keypair::from_bytes(&kp.to_bytes()).expect("failed to copy keypair")
+}
+
+async fn setup() -> (
+    Arc<Mutex<ProgramTestContext>>,
+    Arc<dyn ProgramClient<ProgramBanksClientProcessTransaction>>,
+    Arc<Keypair>,
+) {
+    let mut program_test = ProgramTest::new(
+        "spl_token_upgrade",
+        spl_token_upgrade::id(),
+        processor!(spl_token_upgrade::processor::process),
+    );
+
+    program_test.prefer_bpf(false); // simplicity in the build
+
+    program_test.add_program(
+        "spl_token_2022",
+        spl_token_2022::id(),
+        processor!(spl_token_2022::processor::Processor::process),
+    );
+    program_test.add_program(
+        "spl_token",
+        spl_token::id(),
+        processor!(spl_token::processor::Processor::process),
+    );
+
+    let context = program_test.start_with_context().await;
+    let payer = Arc::new(keypair_clone(&context.payer));
+    let context = Arc::new(Mutex::new(context));
+
+    let client: Arc<dyn ProgramClient<ProgramBanksClientProcessTransaction>> =
+        Arc::new(ProgramBanksClient::new_from_context(
+            Arc::clone(&context),
+            ProgramBanksClientProcessTransaction,
+        ));
+    (context, client, payer)
+}
+
+async fn setup_mint<T: SendTransaction>(
+    program_id: &Pubkey,
+    mint_authority: &Pubkey,
+    decimals: u8,
+    payer: Arc<Keypair>,
+    client: Arc<dyn ProgramClient<T>>,
+) -> Token<T> {
+    let mint_account = Keypair::new();
+    let token = Token::new(client, program_id, &mint_account.pubkey(), payer);
+    token
+        .create_mint(mint_authority, None, decimals, vec![], &[&mint_account])
+        .await
+        .unwrap();
+    token
+}
+
+#[test_case(spl_token::id(), spl_token_2022::id() ; "upgrade to token-2022")]
+#[test_case(spl_token_2022::id(), spl_token::id() ; "downgrade to token")]
+#[test_case(spl_token::id(), spl_token::id() ; "token to token")]
+#[test_case(spl_token_2022::id(), spl_token_2022::id() ; "token-2022 to token-2022")]
+#[tokio::test]
+async fn success(original_program_id: Pubkey, new_program_id: Pubkey) {
```

Comment:
this is really cool i had no idea this was a thing

Reply:
hehe it's the cornerstone to testing the upgrade to token-swap